0011-resharding-without-downtime.md

```diff
@@ -0,0 +1,258 @@
+    map<string, Shard> shard_map = 3;
+}
+```
+
+### Other Algorithms
+
+Other algorithms considered were: [Consistent
+hashing](https://en.wikipedia.org/wiki/Consistent_hashing) and
+[Maglev](https://storage.googleapis.com/gweb-research2023-media/pubtools/2904.pdf).
+
+Rendezvous hashing was chosen for it's simpleness of implementation,
+mathematical elegance and not needing to be parameterized in configuration.
+
+_Of all shards pick the one with the largest `-weight / log(h)` where h is a
+hash of the key/node pair mapped to `]0, 1[`._
+
+There are some minor advantages and disadvantages to the other algorithms, in
+particular with regards to how computationally heavy they are. However, in a
+typical buildbarn cluster with dozens of shards, the amount of computation
+required to pick a shard is unlikely to be relevant.
+
+##### Consistent Hashing
+
+Consistent hashing can be reduced into a special case of Rendezvous with the
+advantage that it can precomputed into a sorted list. There we can binary search
+for the target in `log(W)` time. This is in contrast to to Rendezvous hashing
+which takes `N` time to search.
+
+* `N` is the number of shards
+* `W` is the sum of weights in its reduced form (i.e. smallest weight is 1 and
+  all weights are integer)
+
+In practice buildbarn clusters only have a small number of shards and doing a
+straight linear search with Rendezvous is faster for any reasonable cluster (say
+less than 100 shards). Rendezvous is also independent on the weight of the
+individual shards which improves the relative performance of Rendezvous for
+shards that are not evenly weighted.
+
+#### Maglev
+
+Maglev in turn uses an iterative method that precomputes an evenly distributed
+lookup table of any arbitrary size. After the precomputational step the target
+shard is a simple lookup of the key modulo the size of the table.
+
+The disadvantage of Maglev is that it aims for minimal resharding rather than
+optimal resharding. When removing a shard from the set of all shard, the
+algorithm tolerates a small amount of unecessary resharding. This is the order
+of a few percentiles depending on the number of shards and size of the lookup
+table.
+
+## Handling Topology Changes Dynamically
+
+To reduce the amount of restarts required buildbarn needs a construct that
+describes how to reload the topology.
+
+### RemotelyDefinedBlobAccessConfiguration
+
+```
+message RemotelyDefinedBlobAccessConfiguration {
+    // Fetch the blob access configuration from an external service
+    buildbarn.configuration.grpc.ClientConfiguration endpoint = 1;
+
+    // Duration to reuse the previous configuration when not able to reach
+    // RemoteBlobAccessConfiguration.GetBlobAccessConfiguration
+    // before the component should consider itself partitioned from the cluster
+    // and return UNAVAILABLE for any access requests.
+    //
+    // Recommended value: 10s
+    google.protobuf.Duration remote_configuration_timeout = 2;
```

Comment:
Have you considered supporting a time-to-live (TTL) value **from** the fetched configuration, telling how many seconds that particular retrieved configuration is valid? Similar to TTL in DNS records. **That would allow changing the refresh interval without restarting all components**. Operators could for example freely choose a propagation duration between step 2 and 3 (in the 'Context' chapter) for each resharding occasion, by first changing the refresh interval (for scenarios where unnecessary client retries is undesired).

If configuration is retrieved via HTTP, then a TTL expire value could be specified in a HTTP response header.